lib/mailjet/mailer.rb

```diff
@@ -27,25 +27,187 @@ def initialize(options = {})
+  def deliver!(mail, options = nil)
+#    p mail.header.fields
+    
+    if (options && options.kind_of?(Object) && options['version'].present?)
+      @version = options['version']
+    end
+    
+    if (!options.kind_of?(Object))
+      options = []
+    end
+
+    # Mailjet Send API does not support full from. Splitting the from field into two: name and email address
+    if mail[:from].nil? && Mailjet.config.default_from.present?
+      mail[:from] = Mailjet.config.default_from
+    end
+
+    if (@version == 'v3.1')
+      Mailjet::Send.create({:Messages => [setContentV3_1(mail)]})
+    else
+      Mailjet::Send.create(setContentV3_0(mail))
+    end
+  end
+
+  def setContentV3_1(mail)
+    content = {}
+    content[:TextPart] = mail.text_part.try(:decoded)
+    content[:HTMLPart] = mail.html_part.try(:decoded)
+    
+    unless mail.attachments.empty?
+      content[:Attachments] = []
+      content[:InlineAttachments] = []
+
+      mail.attachments.each do |attachment|
+        mailjet_attachment = {
+          'ContentType' => attachment.content_type.split(';')[0],
+          'Filename' => attachment.filename,
+          'Base64Content' => Base64.encode64(attachment.body.decoded)
+        }
+
+        if attachment.inline?
+          mailjet_attachment['ContentId'] = attachment.content_id
+          content[:InlineAttachments].push(mailjet_attachment)
+        else
+          content[:Attachments].push(mailjet_attachment)
+        end
+      end
+    end
+    
+    # We do the next part in purpose to accept only custom header from the user, but without accepting the mailjet related ones
+    # The mailjet related ones are stocked in another variable and directly added to the body
+    if mail.header && !mail.header.fields.empty?
+      content[:Headers] = {}
+      mail.header.fields.each do |header|
+        if header.name.start_with?('X-') && !header.name.start_with?('X-MJ') && !header.name.start_with?('X-Mailjet')
+          content[:Headers][header.name] = header.value
+        end
+      end
+    end
+
+    # Reply-To is not a property in Mailjet Send API
+    # Passing it as an header if mail.reply_to
+    
+    if mail.reply_to
+      if mail.reply_to.display_names.first
+        content[:Headers]['Reply-To'] = {:Email=> mail[:reply_to].addresses.first, :Name=> mail[:reply_to].display_names.first}
+      else
+        content[:Headers]['Reply-To'] = {:Email=> mail[:reply_to].addresses.first}
+      end
+    end
+    
+    if (mail[:to])
+      if (mail[:to].is_a? String)
+        if mail[:to].display_names.first
+          to = [{:Email=>mail[:to].addresses.first, :Name=>mail[:to].display_names.first}]
+        else
+          to = [{:Email=>mail[:to].addresses.first}]
+        end
+      else
+        to = []
+        mail[:to].each do |t|
+          if (t.display_name)
+            to << {:Email=> t.address, :Name=>t.display_name}
+          else
+            to << {:Email=> t.address}
+          end
+        end
```

Comment:
using `Array#map`, it could be rewritten by
```ruby
to = mail[:to].map do |t|
  if (t.display_name)
    { :Email => t.address, :Name => t.display_name }
  else
    { :Email => t.address }
  end
end
```